File: Dart/11-Math&Geometry/Problems/4-RotateImage/RotateImage.dart
// Dart program implementing Rotate Image using 5 different approaches

void main() {
  List<List<int>> image = [
    [1, 2, 3],
    [4, 5, 6],
    [7, 8, 9]
  ];

  print("Original Image:");
  printMatrix(image);

  print("\n1. Rotate Using Transpose & Reverse:");
  printMatrix(rotateTransposeReverse(image));

  print("\n2. Rotate Using Extra Matrix:");
  printMatrix(rotateExtraMatrix(image));

  print("\n3. Rotate In-Place (Layer by Layer):");
  printMatrix(rotateInPlace(image));

  print("\n4. Rotate Using Matrix Multiplication:");
  printMatrix(rotateMatrixMultiplication(image));

  print("\n5. Rotate Using Recursive Approach:");
  printMatrix(rotateRecursive(image));
}

// Helper function to print a 2D matrix
void printMatrix(List<List<int>> matrix) {
  for (var row in matrix) {
    print(row);
  }
}

// Approach 1: Rotate using Transpose & Reverse
// Rotates the matrix 90 degrees clockwise by first transposing it and then reversing each row.
List<List<int>> rotateTransposeReverse(List<List<int>> matrix) {
  int n = matrix.length;
  // Transpose the matrix: Swap elements across the main diagonal.
  for (int i = 0; i < n; i++) {
    for (int j = i + 1; j < n; j++) {
      int temp = matrix[i][j]; // Store the element to be swapped.
      matrix[i][j] = matrix[j][i]; // Swap elements.
      matrix[j][i] = temp; // Complete the swap.
    }
  }
  // Reverse each row: Reflect each row about its center.
  for (int i = 0; i < n; i++) {
    matrix[i] = matrix[i].reversed.toList(); // Reverse the elements in the current row.
  }
  return matrix; // Return the rotated matrix.
}

// Approach 2: Rotate Using Extra Matrix
// Rotates the matrix 90 degrees clockwise by creating a new matrix and copying elements in rotated order.
List<List<int>> rotateExtraMatrix(List<List<int>> matrix) {
  int n = matrix.length;
  // Create a new matrix with the same dimensions.
  List<List<int>> result = List.generate(n, (_) => List<int>.filled(n, 0));
  // Copy elements from the original matrix to the new matrix in rotated order.
  for (int i = 0; i < n; i++) {
    for (int j = 0; j < n; j++) {
      result[j][n - i - 1] = matrix[i][j]; // Calculate the new position and copy.
    }
  }
  return result; // Return the rotated matrix.
}

// Approach 3: Rotate In-Place (Layer by Layer)
// Rotates the matrix 90 degrees clockwise in-place by rotating layers of elements.
List<List<int>> rotateInPlace(List<List<int>> matrix) {
  int n = matrix.length;
  // Iterate through the layers of the matrix.
  for (int layer = 0; layer < n ~/ 2; layer++) {
    int first = layer; // Index of the first element in the layer.
    int last = n - layer - 1; // Index of the last element in the layer.
    // Iterate through the elements in the current layer.
    for (int i = first; i < last; i++) {
      int offset = i - first; // Calculate the offset from the first element.
      int top = matrix[first][i]; // Store the top element.
      // Rotate the elements in the layer: left -> top, bottom -> left, right -> bottom, top -> right.
      matrix[first][i] = matrix[last - offset][first];
      matrix[last - offset][first] = matrix[last][last - offset];
      matrix[last][last - offset] = matrix[i][last];
      matrix[i][last] = top;
    }
  }
  return matrix; // Return the rotated matrix.
}

// Approach 4: Rotate Using Matrix Multiplication
// Rotates the matrix 90 degrees clockwise using matrix multiplication principles.
List<List<int>> rotateMatrixMultiplication(List<List<int>> matrix) {
  int n = matrix.length;
  // Rotation matrix for 90 degrees clockwise.
  List<List<int>> rotationMatrix = [
    [0, -1],
    [1, 0]
  ];
  // Create a new matrix to store the rotated result.
  List<List<int>> result = List.generate(n, (_) => List<int>.filled(n, 0));
  // Calculate the new coordinates for each element using matrix multiplication.
  for (int i = 0; i < n; i++) {
    for (int j = 0; j < n; j++) {
      int x = rotationMatrix[0][0] * i + rotationMatrix[0][1] * j + (n - 1); // Calculate new x coordinate.
      int y = rotationMatrix[1][0] * i + rotationMatrix[1][1] * j; // Calculate new y coordinate.
      result[x][y] = matrix[i][j]; // Copy the element to the new position.
    }
  }
  return result; // Return the rotated matrix.
}

// Approach 5: Rotate Using Recursive Approach
// Rotates the matrix 90 degrees clockwise using a recursive approach.
List<List<int>> rotateRecursive(List<List<int>> matrix) {
  // Recursive helper function to rotate submatrices.
  void helper(int row, int col, int n) {
    if (n <= 1) return; // Base case: single element or empty submatrix.
    int temp = matrix[row][col]; // Store the top-left element.
    // Rotate the four corner elements.
    matrix[row][col] = matrix[row + n - 1][col];
    matrix[row + n - 1][col] = matrix[row + n - 1][col + n - 1];
    matrix[row + n - 1][col + n - 1] = matrix[row][col + n - 1];
    matrix[row][col + n - 1] = temp;
    // Recursively rotate the inner submatrix.
    helper(row + 1, col + 1, n - 2);
  }
  helper(0, 0, matrix.length); // Start the recursive rotation from the top-left corner.
  return matrix; // Return the rotated matrix.
}

File: Java/11-Math&Geometry/Problems/4-RotateImage/RotateImage.java
import java.util.Arrays;

public class RotateImage {

    public static void main(String[] args) {
        int[][] matrix = {
                {1, 2, 3},
                {4, 5, 6},
                {7, 8, 9}
        };

        System.out.println("Original Matrix:");
        printMatrix(matrix);

        // Approach 1: Using Extra Space (Creating a new matrix)
        int[][] rotatedMatrix1 = rotateUsingExtraSpace(matrix);
        System.out.println("\nRotated Matrix (Approach 1 - Extra Space):");
        printMatrix(rotatedMatrix1);

        // Approach 2: In-Place Rotation (Optimal)
        int[][] matrix2 = copyMatrix(matrix); // Create a copy to avoid modifying the original
        rotateInPlace(matrix2);
        System.out.println("\nRotated Matrix (Approach 2 - In-Place):");
        printMatrix(matrix2);

        // Approach 3: Layer-by-Layer Rotation
        int[][] matrix3 = copyMatrix(matrix);
        rotateLayerByLayer(matrix3);
        System.out.println("\nRotated Matrix (Approach 3 - Layer by Layer):");
        printMatrix(matrix3);

        // Approach 4: Using Collections (for demonstration)
        int[][] matrix4 = copyMatrix(matrix);
        rotateUsingCollections(matrix4);
        System.out.println("\nRotated Matrix (Approach 4 - Collections):");
        printMatrix(matrix4);

        // Approach 5: Transpose and Reverse
        int[][] matrix5 = copyMatrix(matrix);
        rotateTransposeReverse(matrix5);
        System.out.println("\nRotated Matrix (Approach 5 - Transpose and Reverse):");
        printMatrix(matrix5);
    }

    // Helper method to print the matrix
    public static void printMatrix(int[][] matrix) {
        for (int[] row : matrix) {
            System.out.println(Arrays.toString(row));
        }
    }

    // Helper method to copy the matrix, so original matrix is not changed.
    public static int[][] copyMatrix(int[][] matrix) {
        int rows = matrix.length;
        int cols = matrix[0].length;
        int[][] newMatrix = new int[rows][cols];
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < cols; j++) {
                newMatrix[i][j] = matrix[i][j];
            }
        }
        return newMatrix;
    }

    // Approach 1: Using Extra Space
    // - Create a new matrix with dimensions swapped (n x n becomes n x n, but elements are re-arranged).
    // - Copy elements from the original matrix to the new matrix in the rotated order.
    // - Time Complexity: O(n^2), Space Complexity: O(n^2)
    public static int[][] rotateUsingExtraSpace(int[][] matrix) {
        int n = matrix.length;
        int[][] rotatedMatrix = new int[n][n];
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < n; j++) {
                rotatedMatrix[j][n - 1 - i] = matrix[i][j];
            }
        }
        return rotatedMatrix;
    }

    // Approach 2: In-Place Rotation (Optimal)
    // - Rotate the matrix in-place by considering it in layers/cycles.
    // - For each layer, rotate the four edges one by one.
    // - This is the most efficient approach in terms of space complexity.
    // - Time Complexity: O(n^2), Space Complexity: O(1)
    public static void rotateInPlace(int[][] matrix) {
        int n = matrix.length;
        for (int i = 0; i < (n + 1) / 2; i++) { // Iterate through layers
            for (int j = 0; j < n / 2; j++) {
                int temp = matrix[i][j]; // Store the top element

                // Rotate elements in a cycle
                matrix[i][j] = matrix[n - 1 - j][i];            // left -> top
                matrix[n - 1 - j][i] = matrix[n - 1 - i][n - 1 - j];  // bottom -> left
                matrix[n - 1 - i][n - 1 - j] = matrix[j][n - 1 - i];  // right -> bottom
                matrix[j][n - 1 - i] = temp;                      // top -> right
            }
        }
    }

    // Approach 3: Layer-by-Layer Rotation (Similar to In-Place, but more explicit)
    public static void rotateLayerByLayer(int[][] matrix) {
        int n = matrix.length;
        for (int layer = 0; layer < n / 2; layer++) {
            int first = layer;
            int last = n - 1 - layer;
            for (int i = first; i < last; i++) {
                int offset = i - first;
                int top = matrix[first][i]; //save top

                //left -> top
                matrix[first][i] = matrix[last - offset][first];

                //bottom -> left
                matrix[last - offset][first] = matrix[last][last - offset];

                //right -> bottom
                matrix[last][last - offset] = matrix[i][last];

                //top -> right
                matrix[i][last] = top; //right
            }
        }
    }

    // Approach 4: Using Collections (Demonstration - Not the most efficient)
    // - This approach uses Java Collections to rotate the matrix.
    // - It's less efficient than the in-place methods due to the overhead of using Collections.
    // - Included for demonstration purposes.
    public static void rotateUsingCollections(int[][] matrix) {
        int n = matrix.length;
        // Transpose the matrix (swap rows and columns)
        for (int i = 0; i < n; i++) {
            for (int j = i; j < n; j++) {
                int temp = matrix[i][j];
                matrix[i][j] = matrix[j][i];
                matrix[j][i] = temp;
            }
        }
        // Reverse each row
        for (int i = 0; i < n; i++) {
            reverseArray(matrix[i]);
        }
    }

    // Helper method to reverse an array (used in Approach 4)
    private static void reverseArray(int[] arr) {
        int left = 0;
        int right = arr.length - 1;
        while (left < right) {
            int temp = arr[left];
            arr[left] = arr[right];
            arr[right] = temp;
            left++;
            right--;
        }
    }

    // Approach 5: Transpose and Reverse Rows
    // 1. Transpose the matrix: Swap elements across the main diagonal.
    // 2. Reverse each row of the transposed matrix.
    public static void rotateTransposeReverse(int[][] matrix) {
        int n = matrix.length;

        // Transpose the matrix
        for (int i = 0; i < n; i++) {
            for (int j = i + 1; j < n; j++) { // Start from i+1 to avoid double swapping
                int temp = matrix[i][j];
                matrix[i][j] = matrix[j][i];
                matrix[j][i] = temp;
            }
        }

        // Reverse each row
        for (int i = 0; i < n; i++) {
            int left = 0;
            int right = n - 1;
            while (left < right) {
                int temp = matrix[i][left];
                matrix[i][left] = matrix[i][right];
                matrix[i][right] = temp;
                left++;
                right--;
            }
        }
    }
}

